homeassistant/components/update/__init__.py

```diff
@@ -0,0 +1,257 @@
+            domain=msg["domain"],
+            identifier=msg["identifier"],
+            version=msg["version"],
+            backup=msg.get("backup"),
+        )
+    except Exception as err:  # pylint: disable=broad-except
+        connection.send_error(
+            msg["id"],
+            "update_failed",
+            f"Update of {msg['identifier']} to version {msg['version']} failed: {err}",
+        )
+
+    connection.send_result(msg["id"])
+
+
+class UpdatePlatformProtocol(Protocol):
+    """Define the format that update platforms can have."""
+
+    async def async_list_updates(self, hass: HomeAssistant) -> list[UpdateDescription]:
+        """List all updates available in the integration."""
+
+    async def async_perform_update(
+        self,
+        hass: HomeAssistant,
+        identifier: str,
+        version: str,
+        **kwargs: Any,
+    ) -> None:
+        """Perform an update."""
+
+
+@dataclasses.dataclass()
+class UpdateDescription:
+    """Describe an update update."""
+
+    identifier: str
+    name: str
+    current_version: str
+    available_version: str
+    changelog_content: str | None = None
+    changelog_url: str | None = None
+    icon_url: str | None = None
+    supports_backup: bool = False
+
+
+class UpdateManager:
+    """Update manager for the update integration."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the update manager."""
+        self._hass = hass
+        self._store = storage.Store(
+            hass=hass,
+            version=STORAGE_VERSION,
+            key=DOMAIN,
+        )
+        self._skip: set[str] = set()
+        self._platforms: dict[str, UpdatePlatformProtocol] = {}
+        self._loaded = False
+
+    async def add_platform(
+        self,
+        hass: HomeAssistant,
+        integration_domain: str,
+        platform: UpdatePlatformProtocol,
+    ) -> None:
+        """Add a platform to the update manager."""
+        self._platforms[integration_domain] = platform
+
+    async def _load(self) -> None:
+        """Load platforms and data from storage."""
+        await integration_platform.async_process_integration_platforms(
+            self._hass, DOMAIN, self.add_platform
+        )
+        from_storage = await self._store.async_load()
+        if isinstance(from_storage, dict):
+            self._skip = set(from_storage["skipped"])
+
+        self._loaded = True
+
+    async def gather_updates(self) -> list[dict[str, Any]]:
+        """Gather updates."""
+        if not self._loaded:
+            await self._load()
+
+        updates: dict[str, list[UpdateDescription] | None] = {}
+
+        for domain, update_descriptions in zip(
+            self._platforms,
+            await asyncio.gather(
+                *(
+                    self._get_integration_info(integration_domain, registration)
+                    for integration_domain, registration in self._platforms.items()
+                )
```

Comment:
This can cause exceptions that would end up cancelling the other gathers. You need to return exceptions and log them.

```suggestion
                    for integration_domain, registration in self._platforms.items()
                ),
                return_exceptions=True
```

Reply:
nm, just noticed `_get_integration_info` does this.